### PR TITLE
Search - Highlight newly added expense

### PR DESCRIPTION
**Pull request by ikevin127**
### Fixed Issues

$ https://github.com/Expensify/App/issues/48716
PROPOSAL: 

### Tests

1. Switch to the Search page tab.
2. Click the global create button (FAB) > Submit expense > Submit an expense.
3. Verify you remain on the search page.
4. Verify that the newly added expense appears in the table and the row is briefly highlighted to draw attention to it.

- [x] Verify that no errors appear in the JS console

### Offline tests

1. Switch to the Search page tab.
2. Go offline.
3. Click the global create button (FAB) > Submit expense > Submit an expense.
4. Verify you remain on the search page.
5. Go back online.
6. Verify that the newly added expense appears in the table and the row is briefly highlighted to draw attention to it.

### QA Steps

**TLDR**: Same as Tests.

1. Switch to the Search page tab.
2. Click the global create button (FAB) > Submit expense > Submit an expense.
3. Verify you remain on the search page.
4. Verify that the newly added expense appears in the table and the row is briefly highlighted to draw attention to it.

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/user-attachments/assets/9038252a-e614-44b4-abe6-dfadd3f2c69c

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/user-attachments/assets/4533e6be-1d10-4af2-98dd-42bfd5ef2425

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/user-attachments/assets/828a808a-6a51-4d3e-9e00-5cbcc72d3846

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/user-attachments/assets/10f9bd1f-42f3-4e23-a008-159070056677

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/user-attachments/assets/24217552-4286-484c-9530-73d46bc4e418

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/user-attachments/assets/6ca2b167-9474-4062-aa01-3528c22b5b0c

</details>

### CODE REVIEW

File: src/components/Search/index.tsx [hunk 1]
```diff
@@ -123,9 +133,41 @@ function Search({queryJSON}: SearchProps) {
...
+    useEffect(() => {
+        if (!searchTriggeredRef.current) {
+            return;
+        }
+
+        // Schedule a task to reset the flag after all current interactions have completed
+        const interactionHandle = InteractionManager.runAfterInteractions(() => {
+            searchTriggeredRef.current = false; // Reset the flag after interactions
+        });
+
```
Comment (ikevin127) [inline src/components/Search/index.tsx:166]: @luacmartins Ideally, we would update the `Search` snapshot in BE and pop the newly added transaction in the search list via pusher and we wouldn't have to check transactions / previous transactions in this way - but it's better than performing a full page reload from a UX point of view, given that we want to pop the row in with a nice animation.

**Comment by ikevin127** [comment]
@ishpaul777 I assume it does, coming from the issue - currently we're still polishing the animation and I have to fix the failed checks. Will let you know once this is ready for C+ review 👍

**Comment by ikevin127** [comment]
@luacmartins Not sure why the tests are failing, could it be because of the FE implementation mentioned in https://github.com/Expensify/App/pull/49192?notification_referrer_id=NT_kwDOA116B7QxMjM5OTQyNjM4OTo1NjQ1NzczNQ#discussion_r1759544384 ? 🤔

**Comment by luacmartins** [comment]
Not sure either. They all seem unrelated. Are they consistently failing?

**Comment by ikevin127** [comment]
@luacmartins They seem to be (job 1, 2 and performance) - will try another sync w/ main but if that doesn't do it, not sure what the next steps should be here 🤔

**Comment by dubielzyk-expensify** [comment]
I don't seem to always get the highlight animation when submitting from an empty state:

https://github.com/user-attachments/assets/a3e053c6-ea06-4a0e-8879-10d4ac67f654

I've tried to reproduce it and get the animation once in a while, but not reliably

**Comment by dubielzyk-expensify** [comment]
As for the animation, I tested it locally and it looks really good 😄   Great job! 👏 cc @Expensify/design for viz

**Comment by ikevin127** [comment]
> I don't seem to always get the highlight animation when submitting from an empty state:

@dubielzyk-expensify Thanks for double-checking this, you saved us a regression with this one as I missed checking the case when the list is empty (new account) - which caused the following two issues:
1. When the list is empty (new account), there won't be any previous transactions -> the Search get wouldn't be called with previous logic. ✅ Fixed the logic to account for this!
2. When the list is empty (new account), there won't be any previous transactions -> the `shouldAnimateInHighlight` would be `false` because the `newSearchResultKey` would try to determine if new transaction key exists in the previous transactions list, but because this is empty on a new account -> `newSearchResultKey` would be undefined hence the newly added transaction wouldn't animate in / highlight. ✅ Fixed the logic to account for this!

##

This is how it looks now on a new account when creating the first transaction:

https://github.com/user-attachments/assets/cca8b7d1-9d17-4f5b-96ef-608aa8e7c40e

**Comment by dubielzyk-expensify** [comment]
Boom! That's looking excellent

**Comment by luacmartins** [comment]
Nice! This is looking great. @ikevin127 we have some conflicts. Also is this ready for review?

**Comment by ikevin127** [comment]
@luacmartins Yes, this is ready for review - just fixed the conflict, but the test issues are still standing and I'm not sure what can I do to move forward in this case.

**Comment by ikevin127** [comment]
♻️ I will look into this today, use some AI to try and find a link between my changes and the tests that are failing.

**Comment by ikevin127** [comment]
@ishpaul777 This is ready for review now!

More details about how I fixed the failing tests in https://github.com/Expensify/App/pull/49192#issuecomment-2361883304.

**Comment by ikevin127** [comment]
@luacmartins Need your input on https://github.com/Expensify/App/pull/49192#issuecomment-2364224892 and https://github.com/Expensify/App/pull/49192#discussion_r1769016416 as to whether we want to extend the feature to include these details, given that it wasn't mentioned to me before.

File: src/components/Search/index.tsx [hunk 2]
```diff
@@ -182,6 +216,24 @@ function Search({queryJSON}: SearchProps) {
     }
 
     const searchResults = currentSearchResults?.data ? currentSearchResults : lastSearchResultsRef.current;
+    const previousSearchResults = usePrevious(searchResults?.data);
+
+    const newSearchResultKey = useMemo(() => {
```
[the remark this answers is not in the document]
Comment (luacmartins) [reply src/components/Search/index.tsx:221]: Yea, I think that's fine.

**Comment by luacmartins** [comment]
>NAB, Any plans to scroll to expense added when its out of visible view?

yes, we should do this for new expenses

>there can be multiple newsearch keys, example scenario user add few expenses offline then add a new expense online in this case there will be multiple new expenses but only one highlights

This is ok, we need to highlight the latest created transaction

>Currently when you start a chat, you are redirected to the report page (central pane) - instead of staying on the Search page like it happens for Expenses / Invoices, so I don't think that's possible at this point. But if that ever happens, the changes made to TransactionListItem.tsx in this PR covers the Chats component too when it comes to the animation / highlight.

I don't think we need to highlight it in this case, since the RHP would still be open and the new message would be shown to the user. Maybe we can refresh the search page when the RHP is dismissed in this case so the new messages show up.

**Comment by ikevin127** [comment]
>> NAB, Any plans to scroll to expense added when its out of visible view?

> yes, we should do this for new expenses

♻️ Just pushed the **scroll to newly added expense** functionality which works for both Expenses and Invoices.

https://github.com/user-attachments/assets/2f57df12-f792-4b5c-aa90-13a4e7a27543

cc @luacmartins @ishpaul777

**Review by ishpaul777 — APPROVED** [review]
LGTM!

File: src/components/Search/index.tsx [hunk 3]
```diff
@@ -123,9 +132,34 @@ function Search({queryJSON}: SearchProps) {
         }
 
         SearchActions.search({queryJSON, offset});
-        // eslint-disable-next-line react-compiler/react-compiler, react-hooks/exhaustive-deps
     }, [isOffline, offset, queryJSON]);
 
+    const searchTriggeredRef = useRef(false);
```
Comment (luacmartins) [inline src/components/Search/index.tsx:137]: Request: can we move this animation/scrolling logic into a custom hook so we can keep the logic in the Search component small?

**Review by luacmartins — CHANGES_REQUESTED** [review]
BUG: item is not animated when viewing other statuses, e.g. outstanding

https://github.com/user-attachments/assets/5abdf729-0ce4-4245-b4ff-7d40cab52d12

**Comment by ikevin127** [comment]
Thanks for double checking the functionality, I'll start working on fixing the bugs and implementing the hook request tomorrow.

**Comment by ikevin127** [comment]
@luacmartins Thanks for double checking this and catching the bugs 👍 I just applied the fixes for the two bugs mentioned ([ref1](https://github.com/Expensify/App/pull/49192#pullrequestreview-2326671948), [ref2](https://github.com/Expensify/App/pull/49192#issuecomment-2372579039)) and also performed the refactoring requested in https://github.com/Expensify/App/pull/49192#discussion_r1774199378 which ensures we keep the logic in the Search component small.

Here's a video after the fix was applied:

https://github.com/user-attachments/assets/fd08ce05-72fc-4bdf-8849-a7985e39ad8e

**Please check this again and let me know if there are any other issues 🤞**

**Comment by luacmartins** [comment]
Nice! Thanks for working on those. It seems like a few tests are failing. @ishpaul777 could you please review this PR again once the tests above are fixed?

**Comment by ikevin127** [comment]
♻️ Looking into fixing the remaining bugs today.

**Comment by ikevin127** [comment]
@ishpaul777 ♻️ I refactored the code to use a single hook (`useSearchHighlightAndScroll`) and improved the logic to fix the 3 issues you mentioned above. Please test again and let me know if there are any other issues!

**Review by luacmartins — APPROVED** [review]
LGTM

**Comment by ikevin127** [comment]
♻️ Looking into this as I authored the alleged offending PR, will get back soon with confirmation and PR if it is indeed coming from https://github.com/Expensify/App/pull/49192.

**Comment by ikevin127** [comment]
@Julesssss ✅ I confirm that https://github.com/Expensify/App/issues/50046 is indeed a regression of this PR and I will open a follow-up PR to fix it since this is still in the regression period.

File: src/components/SelectionList/Search/ReportListItem.tsx [hunk 4]
```diff
@@ -77,6 +77,9 @@ function ReportListItem<TItem extends ListItem>({
         styles.overflowHidden,
         item.isSelected && styles.activeComponentBG,
         isFocused && styles.sidebarLinkActive,
+        // Removing some of the styles because they are added to the parent OpacityView via animatedHighlightStyle
+        {backgroundColor: 'unset'},
```
Comment (ikevin127) [inline src/components/SelectionList/Search/ReportListItem.tsx:81]: This style overriding caused the following issue:
- https://github.com/Expensify/App/issues/50046

Which we fixed by moving the style up, under the `overflowHidden` and changing it to `styles.bgTransparent` for compatibility with native.

File: src/components/Search/index.tsx [hunk 5]
```diff
@@ -156,6 +165,14 @@ function Search({queryJSON, onSearchListScroll, contentContainerStyle}: SearchPr
 
     const searchResults = currentSearchResults?.data ? currentSearchResults : lastSearchResultsRef.current;
 
+    const {newSearchResultKey, handleSelectionListScroll} = useSearchHighlightAndScroll({
+        searchResults,
+        transactions,
+        previousTransactions,
+        queryJSON,
+        offset,
```
Comment (allroundexperts) [inline src/components/Search/index.tsx:173]: We probably should have added a `offsetForRecentChat` here as well which would have been always `0`. This later caused https://github.com/Expensify/App/issues/54785
